datasimulator/node.py

```diff
@@ -0,0 +1,340 @@
+                )
+        n_samples = min(min_required_samples, n_samples)
+
+        simulated_data = []
+
+        # construct template
+        template = self.construct_property_generator_template(
+            required_only=required_only
+        )
+
+        simulated_data = []
+        for _ in xrange(n_samples):
+            example = {}
+
+            for prop, simple_schema in template.iteritems():
+                if simple_schema["data_type"] is None:
+                    logger.warn(simple_schema["error_msg"])
+                # Skip. Simulate link properties later
+                elif simple_schema["data_type"] == "link_type":
+                    continue
+                else:
+                    example[prop] = Node._simulate_data_from_simple_schema(
+                        prop, simple_schema
+                    )
+
+            example["submitter_id"] = self._simulate_submitter_id()
+            example["type"] = self.name
+
+            simulated_data.append(example)
+
+        # simulate link properties
+        try:
+            self._simulate_link_properties(simulated_data, random)
+            # store to dataset
+            self.simulated_dataset = simulated_data
+        except IndexError:
+            # just skip it
+            pass
+
+        return simulated_data
+
+    def _simulate_link_properties(self, simulated_data, random=False):
+        """
+        Simulate data for required links
+
+        Args:
+            simulated_data(list): list of data samples need to be filled
+            random(bool): whether randomly link to parent nodes
+
+        Outputs:
+            None
+
+        Side effects:
+            simulated_data(list): simulated_data is updated
+
+        """
+        for link_node in self.required_links:
+            for idx, sample in enumerate(simulated_data):
+                if link_node["name"] == "projects":
+                    sample[link_node["name"]] = {"code": self.project}
+                    continue
+
+                if link_node["multiplicity"] in {"many_to_one", "many_to_many"}:
+                    if random:
+                        choosen_sample = random_choice(
+                            link_node["node"].simulated_dataset
+                        )
+                    else:
+                        choosen_sample = link_node["node"].simulated_dataset[idx]
+
+                    if choosen_sample:
+                        sample[link_node["name"]] = {
+                            "submitter_id": choosen_sample["submitter_id"]
+                        }
+                else:
+                    sample[link_node["name"]] = {
+                        "submitter_id": link_node["node"].simulated_dataset[idx][
+                            "submitter_id"
+                        ]
+                    }
+
+    def _simulate_submitter_id(self):
+        return self.name + "_" + generate_string_data()
+
+    @classmethod
+    def simulate_properties_path(cls):
```

Comment:
what is this for?

Reply:
Removed in e6bd4c559e8de5f1b2fe8b05faa47c8103c54e6f